bot/xchange/Cryptopia.php

```diff
@@ -0,0 +1,282 @@
+        }
+
+        $this->pairs = $pairs;
+        $this->confirmationTimes = $conf;
+        $this->tradeFees = $tradeFees;
+        $this->precisions = $precisions;
+        $this->limits = $limits;
+        $this->withdrawFees = $fees;
+        $this->depositFees = $fees;
+
+        $this->calculateTradeablePairs();
+
+    }
+
+    public function getTickers( $currency ) {
+
+        $currency = $this->coinNames[ $currency ];
+        $markets = $this->exchange->loadMarkets();
+
+        $ticker = [ ];
+        foreach ( $markets as $market ) {
+            if ( $market[ 'quote' ] != $currency ) {
+                continue;
+            }
+
+            $ticker[] = $market[ 'symbol' ];
+        }
+
+        $tickers = $this->exchange->fetchTickers( $ticker );
+        $ticker = [ ];
+        foreach ( $tickers as $row ) {
+            $split = explode( '/', $row[ 'symbol' ] );
+
+            // The API doesn't provide the last price, so we just take an average into our spread. :-(
+            $ticker[ strtoupper( $split[ 0 ] ) ] = formatBTC( $row[ 'last' ] );
+        }
+
+        return $ticker;
+
+    }
+
+    public function withdraw( $coin, $amount, $address, $tag = null ) {
+
+        $coin = $this->coinNames[ $coin ];
+        if ( floatval( $amount ) === 0 ) {
+            // Simulate an error which ccxt may not raise for us
+            return false;
+        }
+
+        return $this->checkAPIReturnValue( $this->exchange->withdraw( $coin, formatBTC( $amount ), $address, $tag ) );
+
+    }
+
+    public function getDepositAddress( $coin ) {
+
+        $coin = $this->coinNames[ $coin ];
+        $currencies = $this->exchange->fetch_currencies();
+
+        if ( !key_exists( $coin, $currencies ) || !$currencies[ $coin ] ) {
+            logg( $this->prefix() . "Coin '$coin' is disabled by Exchange!", true );
+            return null;
+        }
+
+        $result = $this->exchange->fetch_deposit_address( $coin );
+        if ( !isset( $result[ 'address' ] ) ) {
+            logg( $this->prefix() . "Please generate a deposit address for $coin!", true );
+            return null;
+        }
+        if ( key_exists( 'tag', $result ) && !is_null( $result[ 'tag' ] ) ) {
+            return array( $result[ 'address' ], $result[ 'tag' ] );
+        }
+        return $result[ 'address' ];
+
+    }
+
+    public function buy( $tradeable, $currency, $rate, $amount ) {
+        try {
+            $precisions = $this->getPrecision( $tradeable, $currency );
+            $limits = $this->getLimits( $tradeable, $currency );
+            $fee = $this->tradeFees[ $tradeable . "_" . $currency ];
+            if($limits[ 'amount' ][ 'min' ] > $amount){
+                logg( $this->prefix() . "Amount for $tradeable less MIN in buy()!" );
+                return null;
+            }
+            $amount_fee = round($amount * $fee, $precisions[ 'amount' ]);
+            print( $this->prefix() . "INFO BUY [$tradeable]:\n" );
```

Comment:
Can you please remove the debugging statements from your code before submitting?  Thanks!